### PR TITLE
Enabled CSVDiff to use gold_dir and added tests

### LINKED ISSUE #10647: 'gold_dir' not respected for CSVDiff
## Rationale
The `gold_dir` option for `CSVDiff` tests doesn't seem to be respected. `Exodiff` works great, with similar input.

## Description
If you move the `kernels/ode` gold directory back one directory:
```
mv test/tests/kernels/ode/gold test/tests/kernels/gold
```
and then add `gold_dir = '../gold'` to each of the tests in `test/tests/kernels/ode`, the `Exodiff` tests succeed, but the `CSVDiff` fails.

## Impact
This will allow `CSVDiff` to use gold directories other than the default.

### PR DESCRIPTION
**Pull request by tophmatthews**
Closes #10647

### CODE REVIEW

**Comment by tophmatthews** [comment]
Whoa, the testing for the TestHarness is complicated...but I think I have the necessary pieces in there to test the `gold_dir` command for `CSVDiff` and `Exodiff`

File: test/tests/test_harness/diff_gold/csvdiff_diff_gold_out.csv [hunk 1]
```diff
@@ -0,0 +1,7 @@
+time,num_dofs
```
Comment (permcody) [inline test/tests/test_harness/diff_gold/csvdiff_diff_gold_out.csv]: Git supports soft links. I'd recommend that you remove these copies and commit soft links (e.g. "ln -s <orig> <link>") instead.

File: python/TestHarness/CSVDiffer.py [hunk 2]
```diff
@@ -10,7 +10,7 @@
 import os, re, math
 
 class CSVDiffer:
-    def __init__(self, test_dir, out_files, abs_zero=1e-11, relative_error=5.5e-6):
+    def __init__(self, test_dir, out_files, abs_zero=1e-11, relative_error=5.5e-6, gold_dir='./gold'):
```
Comment (permcody) [inline python/TestHarness/CSVDiffer.py]: Minor, but we are trying to be better about creating platform neutral python code. There are users that are using this on Windows now. So I'd prefer you to use the "path.join()" utility to create a path without slashes in it for the default above. (os.path.join())

File: test/tests/test_harness/csvdiff_diff_gold.i [hunk 3]
```diff
@@ -0,0 +1,59 @@
+[Mesh]
```
Comment (permcody) [inline test/tests/test_harness/csvdiff_diff_gold.i]: I'm not sure you need to duplicate these files either. I think the right way to test this is to just soft link everything you need or just re-use existing test inputs but specify the different test harness options like you are doing.

**Review by permcody — CHANGES_REQUESTED** [review]
Thanks for the contribution. Can you make a few minor changes before we merge this in?

Thanks!

File: python/TestHarness/CSVDiffer.py (hunk 2 above)
Comment (tophmatthews) [reply python/TestHarness/CSVDiffer.py]: Ah, good to know. I was just duplicating what is in the `./run_tests --dump` comment.

File: test/tests/test_harness/csvdiff_diff_gold.i (hunk 3 above)
Comment (tophmatthews) [reply test/tests/test_harness/csvdiff_diff_gold.i]: yeah, much better with soft links

Comment (tophmatthews) [reply test/tests/test_harness/csvdiff_diff_gold.i]: why are these in the test folder to begin with? To do the diff tests you have to modify
```
test/tests/test_harness/exodiff.i
test/tests/test_harness/diff
python/TestHarness/test_Diff.py
python/TestHarness/tests
```
Seems un-moosian :)

File: test/tests/test_harness/diff_gold/csvdiff_diff_gold_out.csv (hunk 1 above)
Comment (tophmatthews) [reply test/tests/test_harness/diff_gold/csvdiff_diff_gold_out.csv]: much simpler with the ln -s, thanks for the tip!

File: python/TestHarness/CSVDiffer.py (hunk 2 above)
Comment (tophmatthews) [reply python/TestHarness/CSVDiffer.py]: I only removed the `./`, path.join is os.path.join correct?

**Review by permcody — APPROVED** [review]
Very nice! Thanks!

**Pull request merged**